src/RequestParser.cpp

```diff
@@ -0,0 +1,537 @@
+{
+	std::istringstream	requestStream(request);
+	
+	// Step 1: Parse the request-line
+	std::string			requestLine;
+	if (!std::getline(requestStream, requestLine) || requestLine.empty()) {
+		m_errorCode = 400;
+		throw std::runtime_error(ERR_MISS_REQUEST_LINE);
+	}
+	requestLine = parseMethod(requestLine);
+	requestLine = checkForSpace(requestLine);
+	requestLine = parseUri(requestLine);
+	requestLine = checkForSpace(requestLine);
+	requestLine = parseVersion(requestLine);
+	checkForCRLF(requestLine);
+
+    // Step 2: Parse headers
+    std::string headerLine;
+	// The end of the headers section is marked by an empty line (\r\n\r\n).
+    while (std::getline(requestStream, headerLine) && headerLine != "\r" && !headerLine.empty()) {
+		if (headerLine[0] == ' ' || headerLine[0] == '\t') {
+			m_errorCode = 400;
+			throw std::runtime_error(ERR_OBSOLETE_LINE_FOLDING);
+		}
+        std::istringstream headerStream(headerLine);
+        std::string headerName;
+        std::string headerValue;
+        if (std::getline(headerStream, headerName, ':')) {
+			checkHeaderName(headerName);
+			// getline() removes trailing \r\n
+            std::getline(headerStream >> std::ws, headerValue);
+			headerValue = trimTrailingWhiteSpaces(headerValue);
+			checkContentLength(headerName, headerValue);
+            m_request.headers[headerName] = headerValue;
+        }
+    }
+	checkTransferEncoding();
+	if (headerLine != "\r") {
+		m_errorCode = 400;
+		throw std::runtime_error(ERR_MISS_CRLF);
+	}
+
+    // Step 3: Parse body (if any)
+	if (m_request.hasBody) {
+		std::string body;
+		while (std::getline(requestStream, body)) {
+			if (!m_request.body.empty())
+				body += "\n";
+			m_request.body += body;
+		}
+	}
+    return m_request;
+}
+
+/**
+ * @brief Parses the HTTP method from the request line.
+ *
+ * This function extracts the HTTP method from the beginning of the provided request line.
+ * It supports the "GET", "POST", and "DELETE" methods. The method is stored in `m_request.method`,
+ * and a corresponding bit flag is set in `m_requestMethod`. If the method is not recognized,
+ * an error code is set to 501 and a `std::runtime_error` is thrown.
+ *
+ * @param requestLine The request line string from which the method is to be parsed.
+ * @return A substring of `requestLine` starting from the character after the parsed method.
+ * @throws std::runtime_error If the method is not implemented, an error is thrown and the error code is set to 501.
+ */
+std::string	RequestParser::parseMethod(const std::string& requestLine)
+{
+	int	i = -1;
+	while (isalpha(requestLine[++i]))
+		m_request.method.push_back(requestLine[i]);
+
+	if (m_request.method == "GET")
+		m_requestMethod = 1 << 0;
+	else if (m_request.method == "POST")
+		m_requestMethod = 1 << 1;
+	else if (m_request.method == "DELETE")
+		m_requestMethod = 1 << 2;
+	else {
+		m_errorCode = 501;
+		throw std::runtime_error(ERR_METHOD_NOT_IMPLEMENTED);
```

Comment:
It is possible to convey the error code or even the whole request struct (although there would be no need for that; just the error code should suffice), in case the parser throws. An extra exception class inheriting from std::exception can have a member error code which it holds during the throw.

class MyException : public std::exception {
public:
    MyException(const std::string& message, int errorCode)
        : m_message(message), m_errorCode(errorCode) {}

    virtual const char* what() const throw() {
        return m_message.c_str();
    }

    int getErrorCode() const {
        return m_errorCode;
    }

private:
    std::string m_message;
    int m_errorCode;
};

After the catch the response builder can continue to run with the retrieved error code.
I think the throw-catch system is a very neat way to exit the current running process without too much hassle (as with the error returns).

Reply:
This would then look like this(?):

```
try {
 HTTPRequest request = RequestParser::parseHttpRequest(requestString); 
 responseBuilder::buildResponse(request);
 ... // more stuff?
}
catch (const RequestParser::MyException& e) {
 Log << e.what();
 responseBuilder::buildResponse(e.getErrorCode()) // overload for only receiving the errror?
 ... // duplicate more stuff?
}
``` 

If you want to stick with the throws, I would rather 

- add the error_code to the HTTPRequest struct 
- pass it as ref to the parser, and then pass this ref around
- if an error is found, set it in the sruct and throw. 
- the HTTPRequest would stay intact, and ResponseBuilder can pick up on that

```
HTTPRequest http;
try {
 RequestParser::parseHttpRequest(requestString, http); 
}
catch (const RequestParser::MyException& e) {
 Log << e.what();
}
ResponseBuilder::buildResponse(http);
... // more stuff?
``` 

Without throw (:D) :
```
HTTPRequest http = RequestParser::parseHttpRequest(requestString);
ResponseBuilder::buildResponse(http);
... // more stuff?
```

Reply:
Yeah but w/o throwing you would need some elaborate system to return out of every function of the parser and response builder which is not represented in example 3. And just thinking back to minishell grieves my heart...

So yeah, I would go for sth like example 2.

Reply:
ok!

Reply:
Externalised HTTPRequest so it exists outside of RequestParser.